cmd/a2l/a2l.go

```
@@ -68,114 +83,232 @@ func getTreeFromString(a2lString string) (result *a2l.RootNodeType, err error) {
+		if parseError = opt.Unmarshal(buffer.Bytes(), tree); parseError == nil {
+			if serializedTree, err = proto.Marshal(tree); err == nil {
+				for _, chunk = range chunkifyBySize(serializedTree, s.chunkSize) {
+					response.SerializedTreeChunk = chunk
+					err = stream.Send(response)
+				}
+			} else {
+				response.Error = proto.String(fmt.Sprintf("An error occure during serialization of Tree: %v", err))
+				err = stream.Send(response)
+			}
+		} else {
+			errString := parseError.Error()
+			response.Error = &errString
+			err = stream.Send(response)
+		}
 	}
 
-	return result, err
+	return err
 }
 
-func (s *grpcA2LImplType) GetA2LFromTree(_ context.Context, request *a2l.A2LFromTreeRequest) (result *a2l.A2LResponse, err error) {
+func (s *grpcA2LImplType) GetA2LFromTree(stream a2l.A2L_GetA2LFromTreeServer) (err error) {
+	var buffer bytes.Buffer
+	var request *a2l.A2LFromTreeRequest
+	var chunk []byte
+	var a2lDataBytes []byte
+	tree := &a2l.RootNodeType{}
+	response := &a2l.A2LResponse{}
 	indent := ""
 	sorted := false
+	// Note: optionsParsed := false avoid to parse option for each chunk
+	optionsParsed := false
 
-	if request.Indent != nil {
-		for i := uint32(0); i < *request.Indent; i++ {
-			indent += " "
+	for {
+		request, err = stream.Recv()
+		if err != nil {
+			if err == io.EOF {
+				err = nil
+			}
+			break
+		}
+		if !optionsParsed {
+			if request.Indent != nil {
+				for i := uint32(0); i < *request.Indent; i++ {
+					indent += " "
+				}
+			}
+			if request.Sorted != nil {
+				sorted = *request.Sorted
+			}
+			optionsParsed = true
 		}
+		// Use a buffer to receive all chunks
+		buffer.Write(request.Tree)
 	}
-
-	if request.Sorted != nil {
-		sorted = *request.Sorted
+	if err == nil {
+		if err = proto.Unmarshal(buffer.Bytes(), tree); err == nil {
+			a2lDataBytes = []byte(tree.MarshalA2L(0, indent, sorted))
+			for _, chunk = range chunkifyBySize(a2lDataBytes, s.chunkSize) {
+				response.A2L = chunk
+				err = stream.Send(response)
```

The `err` value is silently ignored here, right?

---

Because we return directly this err variable